Applications/VpView/vpVdfTrackIO.cxx

```diff
@@ -0,0 +1,266 @@
+        return false;
+        }
+      }
+    }
+
+
+  if (!reader.hasData())
+    {
+    // No data was obtained; die
+    return false;
+    }
+
+  const auto& inTracks = reader.tracks();
+  foreach (const auto& ti, qtEnumerate(inTracks))
+    {
+    const vdfTrackReader::Track& in = ti.value();
+
+    auto* const track = vtkVgTrack::New();
+    track->InterpolateMissingPointsOnInsertOn();
+    track->SetPoints(this->TrackModel->GetPoints());
+
+    // Set track model identifier
+    const auto vtkId = static_cast<vtkIdType>(ti.key().SerialNumber);
+    if (this->TrackModel->GetTrack(vtkId))
+      {
+      const auto fallbackId = this->TrackModel->GetNextAvailableId();
+      std::cout << "Track id " << vtkId
+                << " is not unique: changing id of imported track to "
+                << fallbackId << std::endl;
+      track->SetId(fallbackId);
+      }
+    else
+      {
+      track->SetId(vtkId);
+      }
+
+    // Set track type classifiers
+    if (!in.Classification.empty())
+      {
+      std::map<int, double> toc;
+      foreach (const auto& c, in.Classification)
+        {
+        toc.emplace(this->GetTrackTypeIndex(c.first.c_str()), c.second);
+        }
+      track->SetTOC(toc);
+      }
+
+    // Iterate over track states
+    foreach (const auto& s, in.Trajectory)
+      {
+      // Get frame metadata
+      QScopedPointer<vpFrame> frame;
+      if (this->FrameMap)
+        {
+        frame.reset(new vpFrame);
+        if (!this->FrameMap->getFrame(s.TimeStamp.FrameNumber, *frame))
+          {
+          frame.reset();
+          }
+        }
+
+      // Extract time stamp
+      auto ts = vtkVgTimeStamp{s.TimeStamp};
+      if (!ts.HasTime() && frame)
+        {
+        ts = frame->Time;
+        }
+
+      // Extract point location
+      double p[4] = {s.ImagePoint.X, s.ImagePoint.Y, 0.0, 1.0};
+      if (this->StorageMode == TSM_InvertedImageCoords)
+        {
+        p[1] = this->GetImageHeight() - p[1];
+        }
+      else if (this->StorageMode == TSM_HomographyTransformedImageCoords)
+        {
+        if (!frame || !frame->Homography)
+          {
+          std::cerr << "ERROR: Homgraphy for frame " << ts.GetFrameNumber()
+                    << "is unavailable.  Track point not added!\n";
+          continue;
+          }
+        else
+          {
+          frame->Homography->MultiplyPoint(p, p);
+          p[0] /= p[3];
+          p[1] /= p[3];
+          }
+        }
+
+      // Extract shell points
+      std::vector<float> shell;
+      if (s.ImageObject.empty())
+        {
+        shell.reserve(12);
+        const auto& tl = s.ImageBox.TopLeft;
+        const auto& br = s.ImageBox.BottomRight;
+        addPoint(shell, tl.X, tl.Y);
+        addPoint(shell, tl.X, br.Y);
+        addPoint(shell, br.X, br.Y);
+        addPoint(shell, br.X, tl.Y);
+        }
+      else
+        {
+        shell.reserve(3 * s.ImageObject.size());
+        foreach (const auto& p, s.ImageObject)
```

Comment:
"reuse" of 'p' (local variable to foreach loop) here is potentially confusing - best if different variable name used here or for the point location that is passed to the `InsertNextPoint `call.

Reply:
`sp` it is :slightly_smiling_face:

Reply:
works for me